### ISSUE TITLE
Attempted write to FOLLOWER failed

### ISSUE THREAD

**Issue by dataskills** [neo4j-php/neo4j-php-client#62]
I successfully testing migrating data to a free tier Neo4j Aura instance. I upgraded to the paid tier and tried to run the same migration when I got the following error:
```
Exception 'Laudis\Neo4j\Exception\Neo4jException' with message 'Neo4j errors detected.
First one with code "" and message "No write operations are allowed directly on this database. 
Writes must pass through the leader. The role of this server is: FOLLOWER (Neo.ClientError.Cluster.NotALeader)"'

in /home/406801.cloudwaysapps.com/ehwecrgftx/public_html/vendor/laudis/neo4j-php-client/src/Bolt/BoltUnmanagedTransaction.php:107
```

I was using `neo4j+s` protocol.

Is this a potential problem with the driver or should I be looking elsewhere for a solution?

**Comment by transistive** [comment]
Hello @dataskills,

We also tested the free tier and the driver works. The error is especially strange as the free tier only has one server. Can you tell me a bit more about the server and database configuration, the way you build the client and what version of the driver you are using?

That way I can try and reproduce it.


Kind regards,

Ghlen

**Comment by dataskills** [comment]
I am using the lowest paid tier for Aura:
<img width="1165" alt="Screen Shot 2021-07-07 at 11 30 44 AM" src="https://user-images.githubusercontent.com/4392698/124811194-0ec58680-df17-11eb-93ee-db94b9340223.png">

I don't think there are any configuration options. The only "settings" you can manipulate is the upgrade button.

I am using neo4j-php/neo4j-php-client version 2.0.4.

I am using the client in this way:

```
$graphClient = ClientBuilder::create()
              ->withDriver(
                'neo4j',
                'neo4j+s://myUsername:myPassword@xxxxx.databases.neo4j.io'
              )->build();

$tsx = $graphClient->beginTransaction();

$statements = [];
// $rows is just data rows from a ledger database
 foreach ($rows as $row) {
     $statements[] = Statement::create('CREATE (n:User {userId: ' . $row['userId'] . ', email: "' . $row['username'] . '" })');
}

$tsx->commit($statements);
```

Let me know if I can provide anything else.

**Comment by dataskills** [comment]
I have a ticket open with Aura, I asked if there might be anything on their side, they said: 

> Are you currently using the bolt:// to connect to the Aura instance. If that case, can you please try to use neo4j+s:// protocol to connect to the Aura instance you are using which is routing aware.

**Comment by transistive** [comment]
Hmm, very strange. I tested it with your exact configuration without any problems. What is the name of your database? If it deviates from the default (neo4j) it might be required to change it.

`'neo4j+s://myUsername:myPassword@xxxxx.databases.neo4j.io?database=myDatabase'`.

Good luck!

**Comment by dataskills** [comment]
After further testing I am getting some strange results. I have the free tier instance, and two paid tier instances that I am comparing to each other. I am running the exact same operations on each of them. I am only running `CREATE` statements.

Free Tier: 
- Transactions: everything works
- Auto-committed queries: everything works
- Never received the FOLLOWER error

Lowest Paid Tier 1/2:  
- Transactions: don't work
- Auto-committed queries: **one** worked out of many many tries
- Besides the one success always receive the FOLLOWER error

Lowest Paid Tier 2/2, should be identical to above: 
- Transactions: work, but I have to wait a long time between queries, otherwise I get FOLLOWER error
- Auto-committed queries: work, but I have to wait a long time between queries, otherwise I get FOLLOWER error
- FOLLOWER error if I don't wait a long time between queries

In summary, the free tier works all the time, the other ones can work but only if I wait a super long time between queries. Has anyone tried to: 

- run several statements at once? 
- several operations in a row? 

Again, if there is a better host please let me know.

**Comment by transistive** [comment]
Hello @dataskills,

Thank you for the explanation. Sadly neo4j was unable to send me a paid connection. Would it be possible for you to set up a google meet call? I would love to be able to reproduce it with you. It seems to me to be the only way to fix this.

Please send me an email to ghlen@pm.me

Kind regards,

Ghlen

**Comment by transistive** [comment]
Hello @dataskills,

I have thoroughly reviewed the example you emailed me. I know for a fact the neo4j aura deployment is incredibly complex and works as a black box. This means that routing is complicated as certain errors (such as not a leader) happen frequently.

I would have never guessed these errors would happen every second, but here we are :shrug: I have contacted the neo4j to double-check the incidence rate. It might still be a bug in the driver.

In the meantime transaction functions are your friend! I have developed these exactly for these more complex cases. It will retry the function in case of these errors. For your example a transaction function looks like this:

```php
$client->writeTransaction(static function (TransactionInterface $tsx) use ($rows)
    foreach ($rows as $row) {
        $tsx->run('MERGE (n:User {userId: $ixUser, email: $email })', $row);
     }
});
```

Using this example, I was able to run the queries without any errors 30 times in a row. I will keep this issue open until I went through this with the neo4j team.

Good luck!

**Comment by dataskills** [comment]
Thank you for your help and effort. In preliminary tests I am finding that the solution you provided works. The performance is not great but I have no idea what could be at fault. Please do report back on your findings after you talk to the neo4j team.

**Comment by dataskills** [comment]
I was able to migrate all of the test data to the server using the `writeTransaction` method, but the performance is very poor. I suspected it was slow after my initial tests but creating 4000+ records took hours where it takes a couple of minutes locally. I hope a solution can be found, I will have to look for another way to host the graph, at least for the time being. Let me know if you think it could help to open a new issue about the performance problem.

**Comment by transistive** [comment]
Hello @dataskills,

I have to thank you once again for your patience and persistence. I was finally able to reproduce the bug and fix it. The issue was once again SSL related. Apparently, the free tier and the pro tier require different configuration methods, which I implemented in version 2.0.5.

Please feel free to reopen the issue if it persists in version 2.0.5.

Happy graphing!

**Comment by dataskills** [comment]
I ran into a new issue, I could not reopen this one so I created a new. Again, I appreciate your work and determination.
